app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java

```diff
@@ -1,17 +1,104 @@
-// package ch.epfl.favo.common;
-//
-// import com.google.firebase.firestore.FirebaseFirestore;
-//
-// public final class DatabaseWrapper {
-//
-//    private static FirebaseFirestore mFirestore;
-//
-//    public DatabaseWrapper() {
-//        FirebaseFirestore.setLoggingEnabled(true);
-//        mFirestore = FirebaseFirestore.getInstance();
-//    }
-//
-//    public static FirebaseFirestore getDatabaseInstance() {
-//        return mFirestore;
-//    }
-// }
+package ch.epfl.favo.common;
+
+import android.util.Log;
+
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.FirebaseFirestoreSettings;
+
+import java.util.ArrayList;
+import java.util.Map;
+
+public class DatabaseWrapper<T extends DatabaseDocument> {
+
+    private static final String TAG = "DatabaseWrapper";
+
+    private static FirebaseFirestore firestore;
+
+    private Class<T> cls;
+
+    private static String collection;
+
+    public DatabaseWrapper(String collection, Class<T> cls) {
+        this.cls = cls;
+        this.collection = collection;
+
+        FirebaseFirestore.setLoggingEnabled(true);
+        FirebaseFirestoreSettings settings =
+                new FirebaseFirestoreSettings.Builder().setPersistenceEnabled(true).build();
+
+        this.firestore = FirebaseFirestore.getInstance();
+        this.firestore.setFirestoreSettings(settings);
+    }
+
+    public T getDocument(String documentId) throws RuntimeException {
+        // HACK (arosu): allows user in the scope of the listener to be saved out of scope
+        final ArrayList<T> document = new ArrayList<>();
+        firestore
+                .collection(collection)
+                .document(documentId)
+                .get()
+                .addOnCompleteListener(
+                        task -> {
+                            if (task.isSuccessful()) {
+                                DocumentSnapshot documentSnapshot = task.getResult();
+                                if (documentSnapshot.exists()) {
+                                    Log.d(
+                                            TAG, String.format("Document snapshot data: %s", documentSnapshot.getData()));
+                                    document.add(documentSnapshot.toObject(this.cls));
```

Comment:
This is a bad idea: you are mixing asynchronous queries with synchronous calls. There is no guarantee that `document.add(documentSnapshot.toObject(this.cls))` finishes before `return document.get(0)`, in fact your code could crash non-deterministically at the return call.

Instead of decomposing the Task, you should simply transform it into an async result of your own types (by using a CompletableFuture for instance https://developer.android.com/reference/java/util/concurrent/CompletableFuture) and then handle it at the call-site. In any case, there is no way to handle asynchronous data synchronously (except by blocking a thread, which would defeat the purpose of an async computation and lag your app).

This would also reduce the amount of lines that is not tested (and you can write a TaskToFutureAdapter in a different class, which you can easily test separately).